packages/grid/_modules_/grid/hooks/features/treeData/useGridTreeData.ts

```diff
@@ -0,0 +1,149 @@
+import { useFirstRender } from '../../utils/useFirstRender';
+import { buildRowTree } from '../../../utils/rowTreeUtils';
+import { GridRowGroupingPreProcessing } from '../../core/rowGroupsPerProcessing';
+import { isFunction } from '../../../utils/utils';
+import { gridFilteredDescendantCountLookupSelector } from '../filter';
+import { GridPreProcessingGroup } from '../../core/preProcessing';
+
+/**
+ * Only available in DataGridPro
+ * @requires useGridPreProcessing (method)
+ * @requires useGridRowGroupsPreProcessing (method)
+ */
+export const useGridTreeData = (
+  apiRef: GridApiRef,
+  props: Pick<
+    GridComponentProps,
+    'treeData' | 'getTreeDataPath' | 'groupingColDef' | 'defaultGroupingExpansionDepth'
+  >,
+) => {
+  const groupingColDef = React.useMemo<GridColDef>(() => {
+    const propGroupingColDef = props.groupingColDef;
+
+    const baseColDef: GridColDef = {
+      ...GRID_TREE_DATA_GROUP_COL_DEF,
+      headerName: apiRef.current.getLocaleText('treeDataGroupingHeaderName'),
+    };
+    let colDefOverride: Partial<GridColDef>;
+
+    if (isFunction(propGroupingColDef)) {
+      const params: GridColDefOverrideParams = {
+        colDef: baseColDef,
+      };
+
+      colDefOverride = propGroupingColDef(params);
+    } else {
+      colDefOverride = propGroupingColDef ?? {};
+    }
+
+    return {
+      ...baseColDef,
+      ...colDefOverride,
+    };
+  }, [apiRef, props.groupingColDef]);
+
+  const updateColumnsPreProcessing = React.useCallback(() => {
+    const addGroupingColumn = (columns: GridColumns) => {
+      if (!props.treeData) {
+        return columns;
+      }
+
+      const index = columns[0].type === 'checkboxSelection' ? 1 : 0;
+
+      return [...columns.slice(0, index), groupingColDef, ...columns.slice(index)];
+    };
+
+    apiRef.current.unstable_registerPreProcessor(
+      GridPreProcessingGroup.hydrateColumns,
+      'treeData',
+      addGroupingColumn,
+    );
+  }, [apiRef, props.treeData, groupingColDef]);
+
+  const updateRowGrouping = React.useCallback(() => {
+    if (!props.treeData) {
+      return apiRef.current.unstable_registerRowGroupsBuilder('treeData', null);
+    }
+
+    const groupRows: GridRowGroupingPreProcessing = (params) => {
+      if (!props.getTreeDataPath) {
+        throw new Error('MUI: No getTreeDataPath given.');
+      }
+
+      const rows = params.ids
+        .map((rowId) => ({
+          id: rowId,
+          path: props.getTreeDataPath!(params.idRowsLookup[rowId]),
+        }))
+        .sort((a, b) => a.path.length - b.path.length);
+
+      return buildRowTree({
+        rows,
+        ...params,
+        defaultGroupingExpansionDepth: props.defaultGroupingExpansionDepth,
+      });
+    };
+
+    return apiRef.current.unstable_registerRowGroupsBuilder('treeData', groupRows);
+  }, [apiRef, props.getTreeDataPath, props.treeData, props.defaultGroupingExpansionDepth]);
+
+  useFirstRender(() => {
+    updateColumnsPreProcessing();
```

Comment:
I created a hook to abstract this logic of registering pre-processors. You can `git checkout` from https://github.com/mui-org/material-ui-x/pull/2946/files#diff-333cd6c7803e1d6c4ea1f4af0205935a3e6768856016f078abe513b6e19399d7.

Reply:
I copy pasted the file :+1: